File: backend/src/models/Route.js
// models/Route.js
const mongoose = require('mongoose');

const coordinateSchema = new mongoose.Schema({
  lat: { type: Number, required: true },
  lng: { type: Number, required: true },
  name: { type: String, required: true }
}, { _id: false });

const routeSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  vehicleId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Vehicle',
    required: true
  },
  origin: {
    type: coordinateSchema,
    required: true
  },
  destination: {
    type: coordinateSchema,
    required: true
  },
  distance: {
    type: Number,
    required: true
  },
  duration: {
    type: Number,
    required: true
  },
  fuelConsumption: {
    type: Number,
    required: true
  },
  routeType: {
    type: String,
    enum: ['optimal', 'fastest'],
    required: true
  },
  waypoints: [{
    type: coordinateSchema
  }],
  polyline: {
    type: String,
    required: true
  },
  trafficLevel: {
    type: String,
    enum: ['low', 'medium', 'high'],
    default: 'medium'
  },
  estimatedCost: {
    type: Number,
    required: true
  },
  weather: {
    condition: String,
    temperature: Number,
    updatedAt: Date
  },
  status: {
    type: String,
    enum: ['active', 'completed', 'cancelled'],
    default: 'active'
  }
}, {
  timestamps: true
});

routeSchema.pre('save', async function(next) {
  try {
    // Si el estimatedCost ya está establecido y fuelConsumption no ha cambiado,
    // no necesitamos recalcularlo
    if (this.estimatedCost && !this.isModified('fuelConsumption')) {
      return next();
    }

    const FuelPrice = mongoose.model('FuelPrice');
    const Vehicle = mongoose.model('Vehicle');
    
    const vehicle = await Vehicle.findById(this.vehicleId);
    if (!vehicle) {
      throw new Error('Vehículo no encontrado');
    }

    const fuelPrice = await FuelPrice.findOne({ 
      fuelType: vehicle.engineType 
    }).sort({ createdAt: -1 });
    
    if (!fuelPrice) {
      throw new Error('Precio de combustible no encontrado');
    }

    this.estimatedCost = this.fuelConsumption * fuelPrice.price;
    next();
  } catch (error) {
    next(error);
  }
});

module.exports = mongoose.model('Route', routeSchema);

File: frontend/lib/models/route.dart
// lib/models/route_model.dart
class Route {
  final String id;
  final String origin;
  final String destination;
  final double distance;
  final double duration;
  final double fuelConsumption;
  final String routeType;
  final String polyline;
  final String trafficLevel;
  final double estimatedCost;
  final DateTime createdAt;

  Route({
    required this.id,
    required this.origin,
    required this.destination,
    required this.distance,
    required this.duration,
    required this.fuelConsumption,
    required this.routeType,
    required this.polyline,
    required this.trafficLevel,
    required this.estimatedCost,
    required this.createdAt,
  });

  factory Route.fromJson(Map<String, dynamic> json) {
    return Route(
      id: json['_id'] ?? '',
      origin: json['origin'] ?? '',
      destination: json['destination'] ?? '',
      distance: json['distance']?.toDouble() ?? 0.0,
      duration: json['duration']?.toDouble() ?? 0.0,
      fuelConsumption: json['fuelConsumption']?.toDouble() ?? 0.0,
      routeType: json['routeType'] ?? '',
      polyline: json['polyline'] ?? '',
      trafficLevel: json['trafficLevel'] ?? 'medium',
      estimatedCost: json['estimatedCost']?.toDouble() ?? 0.0,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'origin': origin,
      'destination': destination,
      'distance': distance,
      'duration': duration,
      'fuelConsumption': fuelConsumption,
      'routeType': routeType,
      'polyline': polyline,
      'trafficLevel': trafficLevel,
      'estimatedCost': estimatedCost,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
